pkg/k8sctrls/manager.go

```diff
@@ -3,76 +3,195 @@ package k8sctrls
+	// when it cannot access the Kubernetes cluster for a short time.
+	//
+	// Quitting is usually not a bad thing, but in order to avoid restarts due to short-time jitter,
+	// we add a simple counting retry here for mitigation.
+	//
+	// The retry mechanism continues trying 10 times, and the retry period is equal to leadership renew timeout.
+	// So, a retry window is 10 times of the retry period, which is 50 seconds by default.
+	// If the manager has been running for more than one retry window,
+	// either restart or not, the retry counter will be reset.
+	var (
+		retryLimit  = 10
+		retryPeriod = *m.options.RetryPeriod
+		retryWindow = time.Duration(retryLimit) * retryPeriod
+
+		retries        int
+		lastStartEpoch time.Time
+	)
+
+	return wait.PollUntilContextCancel(ctx, retryPeriod, true,
+		func(ctx context.Context) (done bool, err error) {
+			lastStartEpoch = time.Now()
+
+			err = m.doStart(ctx, opts.RestConfig, opts.SetupOptions)
+			if err != nil {
+				// Reset retries.
+				if time.Since(lastStartEpoch) >= retryWindow {
+					retries = 0
+				}
+
+				// Retry by error message.
+				switch errMsg := err.Error(); {
+				case strings.Contains(errMsg, "leader election lost"):
+					// Restart from leader election lost, restart.
+					if retries < retryLimit {
+						m.logger.Info("lost leader election, restarting %d times", retryLimit-retries)
+
+						retries++
+						err = nil
+					}
+				case time.Since(lastStartEpoch) <= retryPeriod && strings.Contains(errMsg, "connection refused"):
+					// Connection refused in short time, restart.
+					if retries < retryLimit {
+						m.logger.Infof("cluster connection refused, restarting %d times", retryLimit-retries)
+
+						retries++
+						err = nil
+					}
+				}
+			}
+
+			return
+		},
+	)
+}
+
+func (m *Manager) doStart(ctx context.Context, restConfig *rest.Config, setupOpts SetupOptions) error {
+	// Notify the manager is not ready.
+	m.isReady.Store(false)
 
-	reconcilers, err := m.Setup(ctx, opts.SetupOptions)
+	// Create manager.
+	mgr, err := ctrl.NewManager(restConfig, m.options)
+	if err != nil {
+		return fmt.Errorf("error creating kubernetes controller manager: %w", err)
+	}
+
+	// Setup manager and get controllers.
+	setupOpts.ReconcileHelper = mgr
+
+	controllers, err := m.Setup(ctx, setupOpts)
 	if err != nil {
 		return err
 	}
 
-	for i := 0; i < len(reconcilers); i++ {
-		if err = reconcilers[i].Setup(mgr); err != nil {
+	// Setup controllers.
+	for i := 0; i < len(controllers); i++ {
+		if err = controllers[i].Setup(mgr); err != nil {
 			return fmt.Errorf("error setting up kubernetes controller: %w", err)
 		}
 	}
 
-	return mgr.Start(ctx)
-}
+	// Watch for cache sync.
+	gopool.Go(func() {
+		err := wait.PollUntilContextCancel(ctx, time.Second, true,
+			func(ctx context.Context) (done bool, err error) {
+				return mgr.GetCache().WaitForCacheSync(ctx), ctx.Err()
+			})
+		if err != nil {
+			return
+		}
```

Comment:
Do think we should add a log here for the cache sync error?

Reply:
the cache waiter has an internal logger.

<img width="657" alt="image" src="https://github.com/seal-io/walrus/assets/3394724/b436f0e8-1866-427c-a274-684acb10c16c">

but the --log-verbosity has not affected this, prefer to use another issue to address.

we can trace another log to ensure the cache has be ready.

```
2023-08-31T19:34:25.734+0800	I	k8sctrl	Starting Controller	{"controller": "job", "controllerGroup": "batch", "controllerKind": "Job"}
```